#VRML_SIM R2023b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto"

WorldInfo {
  basicTimeStep 16
}
Viewpoint {
  orientation 0.2601593616689325 -0.35568075742214544 -0.8976682601805761 4.990345913735541
  position -0.6879497470833533 -1.262768209890189 1.0699731419971281
}
TexturedBackground {
}
TexturedBackgroundLight {
}
Floor {
  size 1 1
}
DEF WALL_1 Solid {
  translation 0.5 0 0.05
  children [
    DEF WALL_SHAPE Shape {
      appearance PBRAppearance {
        baseColor 0.12549 0.368627 0.729412
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.02 0.98 0.1
      }
    }
  ]
  boundingObject USE WALL_SHAPE
}
DEF WALL_2 Solid {
  translation -0.5 0 0.05
  children [
    USE WALL_SHAPE
  ]
  name "solid(1)"
  boundingObject USE WALL_SHAPE
}
DEF WALL_3 Solid {
  translation 0 -0.5 0.05
  rotation 0 0 1 1.5708
  children [
    USE WALL_SHAPE
  ]
  name "solid(2)"
  boundingObject USE WALL_SHAPE
}
DEF WALL_4 Solid {
  translation 0 0.5 0.05
  rotation 0 0 1 1.5708
  children [
    USE WALL_SHAPE
  ]
  name "solid(3)"
  boundingObject USE WALL_SHAPE
}
Robot {
  translation -0.385 0.38 0.04
  rotation 0 0 1 0
  children [
    DEF BODY Shape {
      appearance PBRAppearance {
        baseColor 0.917647 0.145098 0.145098
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.2 0.1 0.05
      }
    }
    DEF WHEEL1 HingeJoint {
      jointParameters HingeJointParameters {
        position 707.9180850044814
        axis 0 1 0
        anchor 0.05 0.06 0
      }
      device [
        RotationalMotor {
          name "wheel1"
        }
      ]
      endPoint Solid {
        translation 0.05 0.06 0
        rotation -0.38265870547582737 0.653287524669414 -0.6532899243328905 3.8725265764890144
        children [
          DEF WHEEL Shape {
            appearance PBRAppearance {
              baseColor 0.305882 0.898039 0.25098
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.02
              radius 0.04
              subdivision 24
            }
          }
        ]
        boundingObject USE WHEEL
        physics Physics {
        }
        linearVelocity 0.031142253568670474 -0.005358023217073638 -4.0077940473581927e-10
        angularVelocity -0.49973294323240036 0.86618336296342 -0.40487688788439097
      }
    }
    DEF WHEEL2 HingeJoint {
      jointParameters HingeJointParameters {
        position 319.8948711343375
        axis 0 1 0
        anchor 0.05 -0.06 0
      }
      device [
        RotationalMotor {
          name "wheel2"
        }
      ]
      endPoint Solid {
        translation 0.05 -0.06 0
        rotation -0.9293807046791979 0.26100862561443383 -0.2610095843544104 4.639213874991992
        children [
          USE WHEEL
        ]
        name "solid(1)"
        boundingObject USE WHEEL
        physics Physics {
        }
        linearVelocity -0.010918160011587323 -0.029626934219898978 1.6106052633458317e-09
        angularVelocity 0.4997335297872788 -0.866184341771765 -0.40487704958653226
      }
    }
    DEF WHEEL3 HingeJoint {
      jointParameters HingeJointParameters {
        position 707.9181083649992
        axis 0 1 0
        anchor -0.05 0.06 0
      }
      device [
        RotationalMotor {
          name "wheel3"
        }
      ]
      endPoint Solid {
        translation -0.05 0.06 0
        rotation -0.3826674534018618 0.6532849626255537 -0.6532873622796195 3.872541837593892
        children [
          USE WHEEL
        ]
        name "solid(2)"
        boundingObject USE WHEEL
        physics Physics {
        }
        linearVelocity 0.010918135336110515 0.02969226815892613 -2.6729026970717866e-10
        angularVelocity -0.49973356074077346 0.8661843270490921 -0.40487708448221454
      }
    }
    DEF WHEEL4 HingeJoint {
      jointParameters HingeJointParameters {
        position 319.8948943148472
        axis 0 1 0
        anchor -0.05 -0.06 0
      }
      device [
        RotationalMotor {
          name "wheel4"
        }
      ]
      endPoint Solid {
        translation -0.05 -0.06 0
        rotation -0.9293863427189943 0.2609985876649338 -0.2609995463680387 4.639219925188624
        children [
          USE WHEEL
        ]
        name "solid(3)"
        boundingObject USE WHEEL
        physics Physics {
        }
        linearVelocity -0.031142278328587302 0.005423357241550668 1.7440912229562678e-09
        angularVelocity 0.49973291227226746 -0.8661833776716349 -0.404876852994763
      }
    }
    DEF DS_RIGHT DistanceSensor {
      translation 0.1 -0.03 0
      rotation 0 0 1 0
      type "sonar"
	  lookupTable [ 0     0  0,
      		        4     400  0  ]          
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.184314 0.596078 0.847059
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "ds_right"
    }
    DEF DS_LEFT DistanceSensor {
      translation 0.1 0.03 0
      rotation 0 0 1 0
      type "sonar"
	  lookupTable [ 0     0  0,
      		        4     400  0  ]
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.184314 0.596078 0.847059
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "ds_left"
    }
  ]
  boundingObject USE BODY
  physics Physics {
    density -1
    mass 1
  }
  controller "four_wheels_collision_avoidance"
  linearVelocity -1.2357996662340804e-08 3.2666991917217575e-05 6.716570263854338e-10
  angularVelocity -1.546824500160252e-08 -7.346856544163671e-09 -0.40462983497951677
}
